statesync/reactor.go

```diff
@@ -27,189 +62,299 @@ const (
+// which will be used for cancellations and deadlines when executing the reactor,
+// a p2p Channel used to receive and send messages, and a router that will be
+// used to get updates on peers.
+func NewReactor(
+	logger log.Logger,
+	conn proxy.AppConnSnapshot,
+	connQuery proxy.AppConnQuery,
+	snapshotCh, chunkCh *p2p.Channel,
+	peerUpdates p2p.PeerUpdates,
+	tempDir string,
+) *Reactor {
 	r := &Reactor{
-		conn:      conn,
-		connQuery: connQuery,
+		conn:            conn,
+		connQuery:       connQuery,
+		snapshotCh:      snapshotCh,
+		snapshotChDone:  make(chan bool),
+		chunkCh:         chunkCh,
+		chunkChDone:     make(chan bool),
+		peerUpdates:     peerUpdates,
+		peerUpdatesDone: make(chan bool),
+		tempDir:         tempDir,
 	}
-	r.BaseReactor = *p2p.NewBaseReactor("StateSync", r)
-	return r
-}
 
-// GetChannels implements p2p.Reactor.
-func (r *Reactor) GetChannels() []*p2p.ChannelDescriptor {
-	return []*p2p.ChannelDescriptor{
-		{
-			ID:                  SnapshotChannel,
-			Priority:            3,
-			SendQueueCapacity:   10,
-			RecvMessageCapacity: snapshotMsgSize,
-		},
-		{
-			ID:                  ChunkChannel,
-			Priority:            1,
-			SendQueueCapacity:   4,
-			RecvMessageCapacity: chunkMsgSize,
-		},
-	}
+	r.BaseService = *service.NewBaseService(logger, "StateSync", r)
+	return r
 }
 
-// OnStart implements p2p.Reactor.
+// OnStart starts separate go routines for each p2p Channel and listens for
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed. No error is returned.
 func (r *Reactor) OnStart() error {
+	// Listen for envelopes on the snapshot p2p Channel in a separate go-routine
+	// as to not block or cause IO contention with the chunk p2p Channel. Note,
+	// we do not launch a go-routine to handle individual envelopes as to not
+	// have to deal with bounding workers or pools.
+	go r.processSnapshotCh()
+
+	// Listen for envelopes on the chunk p2p Channel in a separate go-routine
+	// as to not block or cause IO contention with the snapshot p2p Channel. Note,
+	// we do not launch a go-routine to handle individual envelopes as to not
+	// have to deal with bounding workers or pools.
+	go r.processChunkCh()
+
+	go r.processPeerUpdates()
+
 	return nil
 }
 
-// AddPeer implements p2p.Reactor.
-func (r *Reactor) AddPeer(peer p2p.Peer) {
-	r.mtx.RLock()
-	defer r.mtx.RUnlock()
-	if r.syncer != nil {
-		r.syncer.AddPeer(peer)
+// OnStop stops the reactor by signaling to all spawned go routines to exit and
+// blocking until they all exit. Finally, it will close the snapshot and chunk
+// p2p Channels respectively.
+func (r *Reactor) OnStop() {
+	r.snapshotChDone <- true
+	r.chunkChDone <- true
+	r.peerUpdatesDone <- true
+
+	if err := r.snapshotCh.Close(); err != nil {
```

Comment:
why we're closing channels, which were passed to this reactor as arguments? shouldn't this be a caller's responsibility

Reply:
The sender closes channels. In this context, the "real" reactor is the sender (`snapshotCh.Out|Error` and `chunkCh.Out|Error`), so it has the responsibility of closing them. The shim is the sender of the inbound channel(s), so it closes those.

https://go101.org/article/channel-closing.html#:~:text=The%20Channel%20Closing%20Principle,only%20sender%20of%20the%20channel.